server/backend/api/app/crud/crud_user.py

```diff
@@ -246,9 +247,142 @@ def fetch_initial_requester_info(
         select(models.FamUser)
         .options(
             joinedload(models.FamUser.fam_access_control_privileges),
-            joinedload(models.FamUser.fam_user_terms_conditions)
+            joinedload(models.FamUser.fam_user_terms_conditions),
         )
         .filter(models.FamUser.cognito_user_id == cognito_user_id)
     )
     user = db.scalars(q_stm).unique().one_or_none()
     return user
+
+
+def update_user_info_from_idim_source(
+    db: Session, use_pagination: bool, page: int, per_page: int
+) -> schemas.FamUserUpdateResponse:
+    """
+    Go through each user record in the database,
+    update the user information to match the record in IDIM web service,
+    only for IDIR and Business BCeID users, ignore bc service card users
+    """
+    # get a requester from the database
+    requester = (
+        db.query(models.FamUser)
+        .filter(models.FamUser.user_name == "CMENG")
+        .one_or_none()
+    )
+    # setup IDIM web service
+    api_instance_env = (
+        ApiInstanceEnv.PROD if crud_utils.is_on_aws_prod() else ApiInstanceEnv.TEST
+    )
+    idim_proxy_service = IdimProxyService(requester, api_instance_env)
+
+    # grab fam users from user table
+    fam_users = get_users(db)
+    total_users_count = len(fam_users)
+    LOGGER.debug(f"Total number of users: {total_users_count}")
+    if use_pagination:
+        fam_users = (
+            db.query(models.FamUser)
+            .order_by(models.FamUser.user_id.asc())
+            .offset((page - 1) * per_page)
+            .limit(per_page)
+            .all()
+        )
+        LOGGER.debug(
+            f"Updating information for users on page {page}, there are {per_page} users on each page"
+        )
+
+    success_user_list = []
+    failed_user_list = []
+    ignored_user_list = []
+
+    for user in fam_users:
+        try:
+            LOGGER.debug(
+                f"Updating information for user: {user.user_name}, type: {user.user_type_code}, guid: {user.user_guid}"
+            )
+            search_result = None
+            properties_to_update = {}
+
+            if user.user_type_code == UserType.IDIR:
+                # IDIM web service doesn't support search IDIR by user_guid, so we search by userID
+                search_result = idim_proxy_service.search_idir(
+                    schemas.IdimProxySearchParam(**{"userId": user.user_name})
+                )
+
+                if not user.user_guid:
+                    # if user has no user_guid in our database, add it
+                    properties_to_update = {
+                        models.FamUser.user_guid: search_result.get("guid"),
+                    }
```

Comment:
Hi Catherine, wonder maybe we don't need to add this "guid" to properties_to_update just here.

When you later have the `if search_result.get("found")` when search_result is **found**, just add this "guild" to update there, same as "userId" and all other fields?

Reply:
Hi Ian, that's because we shouldn't update user guid if the user already has that value in our database. 

So basically two cases:

- We have user's user_guid in the database, find user by user_guid, if the information mismatch, we update the username and etc. We should never update user_guid, we rely on that
- If we don't have user's user_guid, then we search by user_name, and add the user_guid into the database

For idir user, since we can't search user by user_guid (IDIM doesn't support), we'll search by user_name. The we only add user_guid, not making changes. Cause we'll always rely on user_guid, not user_name

Reply:
Maybe I am confused,
Isn't that if it finds the user by the user_guid, then it does not matter to add this field to the update list or not since it is the same; but if it finds the user by the user_id, then updating to the user_guid is correct since this is the only user that has the user_guid (when empty in our db)?
Unless there could be user(s) that they all have the same user_guid (with the same type)? Then in that case, db probably throw error and your code does log the error for what's wrong with the user update?

Reply:
Resolved confusion with Catherine.